File: 238_Product_of_Array_Except_Self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int i, product = 1, flag = 1, count = 0;
        vector<int> ans;
        for(i=0; i<nums.size(); i++)
        {
            if(nums[i] != 0)
                product *= nums[i];
            else
                count ++;
        }
        if(count>1)
            product = 0;
        for(i=0; i<nums.size(); i++)
        {
            if(!count)
            {
                if(nums[i] != 0)
                    ans.push_back(product/nums[i]);
                else
                    ans.push_back(product);
            }
            else
            {
                if(nums[i] == 0 && count != nums.size())
                    ans.push_back(product);
                else
                    ans.push_back(0);
            }
        }
        return ans;
    }
};

// beats 15.62% runtime and 10.70% memory
// https://leetcode.com/problems/product-of-array-except-self/submissions/1222105766/

File: 238_Product_of_Array_Except_Self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        lpro, rpro = 1, 1
        larr = [0] * len(nums)
        rarr = [0] * len(nums)
        for i in range(len(nums)):
            j = -i-1
            larr[i] = lpro
            rarr[j] = rpro
            lpro *= nums[i]
            rpro *= nums[j]

        for i in range(len(nums)):
            larr[i] = larr[i]*rarr[i]

        return larr

# beats 22.12% runtime and 32.76% memory
# https://leetcode.com/problems/product-of-array-except-self/submissions/1223043536/
